File: src/main/java/com/teamwizardry/librarianlib/api/gui/Key.java
package com.teamwizardry.librarianlib.api.gui;

import java.util.HashMap;
import java.util.Map;

public class Key {

	public final char character;
	public final int keyCode;
	
	private Key(char character, int keyCode) {
		this.character = character;
		this.keyCode = keyCode;
	}
	
	/**
	 * Maps a keycode to a key
	 */
	private static Map<Integer, Key> pool = new HashMap<>();
	
	public static Key get(char character, int keyCode) {
		if(!pool.containsKey(keyCode))
			pool.put(keyCode, new Key(character, keyCode));
		return pool.get(keyCode);
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + keyCode;
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Key other = (Key) obj;
		if (keyCode != other.keyCode)
			return false;
		return true;
	}
	
}

File: src/main/java/com/teamwizardry/librarianlib/client/gui/Key.kt
package com.teamwizardry.librarianlib.client.gui

import java.util.*

class Key private constructor(val character: Char, val keyCode: Int) {

    override fun hashCode(): Int {
        val prime = 31
        var result = 1
        result = prime * result + keyCode
        return result
    }

    override fun equals(obj: Any?): Boolean {
        if (this === obj)
            return true
        if (obj == null)
            return false
        if (javaClass != obj.javaClass)
            return false
        val other = obj as Key
        return keyCode == other.keyCode
    }

    companion object {

        /**
         * Maps a keycode to a key
         */
        private val pool = HashMap<Int, Key>()

        operator fun get(character: Char, keyCode: Int): Key {
            var key = pool[keyCode] ?: Key(character, keyCode)
            if (!pool.containsKey(keyCode))
                pool.put(keyCode, key)
            return key
        }
    }

}
